vrc-get-vpm/src/utils/copy_recursive.rs

```diff
@@ -1,52 +1,44 @@
+use crate::io;
+use crate::io::{DirEntry, IoTrait};
+use futures::prelude::*;
 use std::collections::VecDeque;
-use std::io;
-use std::path::Path;
-use tokio::fs::create_dir_all;
+use std::path::PathBuf;
 
-pub(crate) async fn copy_recursive(src_dir: Box<Path>, dst_dir: Box<Path>) -> io::Result<()> {
+pub(crate) async fn copy_recursive(
+    src_io: &impl IoTrait,
+    src_dir: PathBuf,
+    dst_io: &impl IoTrait,
+    dst_dir: PathBuf,
+) -> io::Result<()> {
     // TODO: parallelize & speedup
     let mut queue = VecDeque::new();
-    queue.push_front((src_dir.into_path_buf(), dst_dir.into_path_buf()));
+    queue.push_front((src_dir, dst_dir));
 
     while let Some((src_dir, dst_dir)) = queue.pop_back() {
-        let mut iter = tokio::fs::read_dir(src_dir).await?;
-        create_dir_all(&dst_dir).await?;
-        while let Some(entry) = iter.next_entry().await? {
+        let mut iter = src_io.read_dir(&src_dir).await?;
+        dst_io.create_dir_all(&dst_dir).await?;
+        while let Some(entry) = iter.try_next().await? {
             let file_type = entry.file_type().await?;
-            let src = entry.path();
+            let src = src_dir.join(entry.file_name());
             let dst = dst_dir.join(entry.file_name());
 
             if file_type.is_symlink() {
```

Comment:
we have to consider: should we support symlink?

Reply:
I think it's reasonable to process symlink as a simple folders / files. in VCC, they are.

Reply:
Removed special handling for symlink